### PR TITLE
[Enhancement] Change submit button to also save serverside

### PR DESCRIPTION
**Pull request by schultek**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Client: I updated multiple integration tests (Jest) related to the features

### Motivation and Context

When clicking submit in the online editor with unsaved changes, a save request is issues before the commit request. This may lead to wrong behaviour in some edge cases with bad / no internet connection.

### Description

The 'save' and 'submit' actions are now handled in one request. The saveChangedFiles request does now optionally also 
commit the changes, depending on a query parameter.

### Steps for Testing

1. Log in to Artemis
2. Navigate to a programming exercise
3. Commit with and without unsaved changes
4. Check that everything works correctly
5. Inspect the network requests to be only one for saving and committing.

**Test also in the instructor editor with all repositories (Test, Solution, Template, ...)**

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/TestRepositoryResource.java [hunk 1]
```diff
@@ -127,4 +134,52 @@ boolean canAccessRepository(Long exerciseId) {
...
+     */
+    @PutMapping("/test-repository/{exerciseId}/files")
+    public ResponseEntity<Map<String, String>> updateTestFiles(@PathVariable("exerciseId") Long exerciseId, @RequestBody List<FileSubmission> submissions,
+            @RequestParam String commit, Principal principal) {
+        ProgrammingExercise exercise = programmingExerciseService.findWithTemplateParticipationAndSolutionParticipationById(exerciseId);
+        String testRepoName = exercise.getProjectKey().toLowerCase() + "-" + RepositoryType.TESTS.getName();
+        if (versionControlService.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.SERVICE_UNAVAILABLE, "VCSNotPresent");
+        }
+        VcsRepositoryUrl testsRepoUrl = versionControlService.get().getCloneRepositoryUrl(exercise.getProjectKey(), testRepoName);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/TestRepositoryResource.java:155]: Test

Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/TestRepositoryResource.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 2]
```diff
@@ -158,12 +145,13 @@ boolean canAccessRepository(Long participationId) throws IllegalArgumentExceptio
...
      * @param participationId id of participation to which the files belong
      * @param submissions     information about the file updates
+     * @param commit          whether to commit after updating the files
      * @param principal       used to check if the user can update the files
      * @return {Map<String, String>} file submissions or the appropriate http error
      */
     @PutMapping(value = "/repository/{participationId}/files")
     public ResponseEntity<Map<String, String>> updateParticipationFiles(@PathVariable("participationId") Long participationId, @RequestBody List<FileSubmission> submissions,
-            Principal principal) {
+            @RequestParam String commit, Principal principal) {
```
[the remark this answers is not in the document]
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:154]: > When I create a new, default programming exercise with code editor enabled and
> 
> * start the exercise
> * open the code editor
> * open the BubbleSort.java file
> * edit the line as shown in the screenshot
> * and press directly "Submit" shortly afterwards,
> 
> I get "0 of 13 passed" after building and testing despite having done a change that makes one test pass.
> 
> When I press Submit again without doing anything else, I get "1 of 13 passed" after the second build finishes.
> 
> image
> 
> I tried it twice to check and it seems well reproducible.

I could not reproduce this on the test server. Could you check again what specific requests get sent to the server when you do it?

**Comment by schultek** [comment]
> When I create a new, default programming exercise with code editor enabled and
> 
> * start the exercise
> * open the code editor
> * open the BubbleSort.java file
> * edit the line as shown in the screenshot
> * and press directly "Submit" shortly afterwards,
> 
> I get "0 of 13 passed" after building and testing despite having done a change that makes one test pass.
> 
> When I press Submit again without doing anything else, I get "1 of 13 passed" after the second build finishes.
> 
> image
> 
> I tried it twice to check and it seems well reproducible.

I could not reproduce this on the test server. Could you check again what specific requests get sent to the server when you do it?

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java]: Changed it to Boolean

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 3]
```diff
@@ -208,87 +200,15 @@ boolean canAccessRepository(Long participationId) throws IllegalArgumentExceptio
...
-        // If updating the file fails due to an IOException, we send an error message for the specific file and try to update the rest
-        Map<String, String> fileSaveResult = new HashMap<>();
-        submissions.forEach((submission) -> {
-            try {
-                fetchAndUpdateFile(submission, repository);
-                fileSaveResult.put(submission.getFileName(), null);
-            }
-            catch (IOException ex) {
-                fileSaveResult.put(submission.getFileName(), ex.getMessage());
+        if ("true".equals(commit)) {
```
[the remark this answers is not in the document]
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/TestRepositoryResource.java [hunk 4]
```diff
@@ -127,4 +132,51 @@ boolean canAccessRepository(Long exerciseId) {
...
+            FileSubmissionError error = new FileSubmissionError(exerciseId, "checkoutConflict");
+            throw new ResponseStatusException(HttpStatus.CONFLICT, error.getMessage(), error);
+        }
+        catch (GitAPIException | InterruptedException ex) {
+            FileSubmissionError error = new FileSubmissionError(exerciseId, "checkoutFailed");
+            throw new ResponseStatusException(HttpStatus.SERVICE_UNAVAILABLE, error.getMessage(), error);
+        }
+        Map<String, String> fileSaveResult = saveFileSubmissions(submissions, repository);
+
+        if ("true".equals(commit)) {
```
[the remark this answers is not in the document]
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/repository/TestRepositoryResource.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryResource.java [hunk 5]
```diff
@@ -302,4 +315,47 @@ abstract Repository getRepository(Long domainId, RepositoryActionType repository
...
+    private void fetchAndUpdateFile(FileSubmission submission, Repository repository) throws IOException {
+        Optional<File> file = gitService.getFileByName(repository, submission.getFileName());
+
+        if (file.isEmpty()) {
+            throw new IOException("File could not be found.");
+        }
+
+        InputStream inputStream = new ByteArrayInputStream(submission.getFileContent().getBytes(StandardCharsets.UTF_8));
+        Files.copy(inputStream, file.get().toPath(), StandardCopyOption.REPLACE_EXISTING);
+        inputStream.close();
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryResource.java:359]: I would suggest to use
```suggestion
        Files.writeString(file.get().toPath(), submission.getFileContent());
```
unless we have a good reason to use copy here (we should add a comment then).

**Review by MaisiKoleni — COMMENTED** [review]
Looks good to me and seems to work reliably now (I don't know why I had issues before).

I only have one small suggestion code-wise.

**Review by MaisiKoleni — APPROVED** [review]

**Review by JonasPetry — APPROVED** [review]

**Pull request merged**